python/ray/data/_internal/logical/rules/operator_fusion.py

```diff
@@ -139,52 +139,43 @@ def _can_fuse(self, down_op: PhysicalOperator, up_op: PhysicalOperator) -> bool:
+                and isinstance(down_logical_op, AbstractMap)
+            )
+            or (
+                isinstance(up_logical_op, AbstractMap)
+                and isinstance(down_logical_op, RandomShuffle)
+            )
+            or (
+                isinstance(up_logical_op, AbstractMap)
+                and isinstance(down_logical_op, Repartition)
+            )
+        ):
             return False
 
         # Do not fuse Repartition operator if shuffle is disabled
         # (i.e. using split shuffle).
         if isinstance(down_logical_op, Repartition) and not down_logical_op._shuffle:
             return False
 
-        # Allow fusing tasks->actors if the resources are compatible (read->map), but
-        # not the other way around. The latter (downstream op) will be used as the
-        # compute if fused.
-        if (
-            isinstance(down_logical_op, AbstractUDFMap)
-            and is_task_compute(down_logical_op._compute)
-            and isinstance(up_logical_op, AbstractUDFMap)
-            and get_compute(up_logical_op._compute)
-            != get_compute(down_logical_op._compute)
-        ):
-            return False
-
-        # Fusing callable classes is only supported if they are the same function AND
-        # their construction arguments are the same. Note the Write can be compatbile
-        # with any UDF as Write itself doesn't have UDF.
-        # TODO(Clark): Support multiple callable classes instantiating in the same actor
-        # worker.
-        if (
-            isinstance(down_logical_op, AbstractUDFMap)
-            and isinstance(down_logical_op._fn, CallableClass)
-            and isinstance(up_logical_op, AbstractUDFMap)
-            and isinstance(up_logical_op._fn, CallableClass)
-            and (
-                up_logical_op._fn != down_logical_op._fn
-                or (
-                    up_logical_op._fn_constructor_args
-                    != down_logical_op._fn_constructor_args
-                    or up_logical_op._fn_constructor_kwargs
-                    != down_logical_op._fn_constructor_kwargs
-                )
-            )
+        if isinstance(down_logical_op, AbstractUDFMap) and isinstance(
+            up_logical_op, AbstractUDFMap
         ):
-            return False
+            # Allow fusing tasks->actors if the resources are compatible (read->map),
+            # but not the other way around. The latter (downstream op) will be used as
+            # the compute if fused.
+            if is_task_compute(down_logical_op._compute) and get_compute(
+                up_logical_op._compute
+            ) != get_compute(down_logical_op._compute):
+                return False
 
         # Only fuse if the ops' remote arguments are compatible.
-        if not _are_remote_args_compatible(
+        if (
+            hasattr(up_logical_op, "_ray_remote_args")
+            and hasattr(down_logical_op, "_ray_remote_args")
```

Comment:
Is this "hasattr" check still needed? I think it was needed because we were fusing limit ops here.